### PR TITLE
Retry db transactions the right way

### PR DESCRIPTION
**Pull request by ScottGarman**
Previously we were retrying db transactions by reusing the same transaction. This doesn't work when CRDB failed the transaction, so ensure we create a new transaction each time we're retrying things.

This also allows the maximum number of retries and the maximum interval between retries to be configured via cmdline/env vars.

### CODE REVIEW

File: internal/upserter/upserter.go [hunk 1]
```diff
@@ -107,6 +109,37 @@ func doUpsert(ctx context.Context, db *sqlx.DB, logger *zap.Logger, id string, i
...
+
+			upsertSuccess = true
+
+			break
+		}
+
+		// Exponential backoff would be overkill here, but adding a bit of jitter
+		// to sleep a short time is reasonable
+		jitter := time.Duration(rand.Int63n(int64(dbRetryInterval)))
+		time.Sleep(jitter)
```
Comment (jordan0day) [inline internal/upserter/upserter.go:131]: should this maybe just go into an `else` block?

File: pkg/api/v1/router_instance_metadata.go [hunk 2]
```diff
@@ -381,6 +383,39 @@ func handleDeleteRequest(c *gin.Context, r *Router, instanceID string, metadata
...
+
+			deleteSuccess = true
+
+			break
+		}
+
+		// Exponential backoff would be overkill here, but adding a bit of jitter
+		// to sleep a short time is reasonable
+		jitter := time.Duration(rand.Int63n(int64(dbRetryInterval)))
+		time.Sleep(jitter)
```
Comment (jordan0day) [inline pkg/api/v1/router_instance_metadata.go:405]: similar comments here as the upserter, around maybe using an `else` block and changing the conditional in the loop a little

File: internal/upserter/upserter.go (hunk 1 above)
Comment (ScottGarman) [reply internal/upserter/upserter.go]: Agree with this recommendation. Updated.

File: internal/upserter/upserter.go [hunk 3]
```diff
@@ -107,6 +109,37 @@ func doUpsert(ctx context.Context, db *sqlx.DB, logger *zap.Logger, id string, i
 		}
 	}
 
+	upsertSuccess := false
+	maxUpsertRetries := viper.GetInt("crdb.max_retries")
+	dbRetryInterval := viper.GetDuration("crdb.retry_interval")
+
+	for i := 0; i <= maxUpsertRetries; i++ {
```
[the remark this answers is not in the document]
Comment (ScottGarman) [reply internal/upserter/upserter.go]: Yeah, I like this better than using `break`s. Updated.

File: pkg/api/v1/router_instance_metadata.go (hunk 2 above)
Comment (ScottGarman) [reply pkg/api/v1/router_instance_metadata.go]: Updated in both the delete retries as well as the upsert retries loops.

**Review by jordan0day — APPROVED** [review]

**Pull request merged**